lib/pf/radius.pm

```diff
@@ -828,96 +830,275 @@ sub switch_access {
+    my $logger = $self->logger;
+    my $merged = { %$options, %$args };
+    $merged->{'rule_class'} = $Rules::AUTH;
+    $merged->{'context'} = $pf::constants::realm::RADIUS_CONTEXT;
+    my $attributes;
+    my $matched = pf::authentication::match2([@$sources], $merged, undef, \$attributes);
+
+    my $values = $matched->{values};
+    $args->{'user_role'} = $values->{$Actions::SET_ROLE};
+
+    $merged->{'rule_class'} = $Rules::ADMIN;
+    $merged->{'context'} = $pf::constants::realm::RADIUS_CONTEXT;
+    $matched = pf::authentication::match2($source_id, $merged, $extra, \$attributes);
+    my $value = $matched->{values}{$Actions::SET_ACCESS_LEVEL} if $matched;
+    if ($value) {
+        my @values = split(',', $value);
+        foreach $value (@values) {
+            if (exists $pf::config::ConfigAdminRoles{$value}->{'ACTIONS'}->{'SWITCH_LOGIN_WRITE'}) {
+                return $args->{'switch'}->returnAuthorizeWrite($args);
+            }
+            if (exists $pf::config::ConfigAdminRoles{$value}->{'ACTIONS'}->{'SWITCH_LOGIN_READ'}) {
+                return $args->{'switch'}->returnAuthorizeRead($args);
+            }
+        }
     } else {
-        my $merged = { %$options, %$args };
-        $merged->{'rule_class'} = $Rules::ADMIN;
-        $merged->{'context'} = $pf::constants::realm::RADIUS_CONTEXT;
-        my $attributes;
-        my $matched = pf::authentication::match2($source_id, $merged, $extra, \$attributes);
-        my $value = $matched->{values}{$Actions::SET_ACCESS_LEVEL} if $matched;
-        if ($value) {
-            my @values = split(',', $value);
-            foreach $value (@values) {
-                if (exists $pf::config::ConfigAdminRoles{$value}->{'ACTIONS'}->{'SWITCH_LOGIN_WRITE'}) {
-                    return $switch->returnAuthorizeWrite($args);
+        $logger->info("User $args->{'user_name'} has no role (Switches CLI - Read or Switches CLI - Write) to permit to login in $args->{'switch'}{'_id'}");
+        return [ $RADIUS::RLM_MODULE_FAIL, ('Reply-Message' => "User has no role defined in PacketFence to allow switch login (SWITCH_LOGIN_READ or SWITCH_LOGIN_WRITE)") ];
+    }
+}
+
+sub mfa_post_auth {
+    my ($self, $args, $options, $sources, $source_id, $extra ,$otp, $password) = @_;
+    my $logger = $self->logger;
+    $logger->info("Pre MFA Authentication");
+    my $merged = { %$options, %$args };
+    $merged->{'rule_class'} = $Rules::AUTH;
+    $merged->{'context'} = $pf::constants::realm::RADIUS_CONTEXT;
+    my $attributes;
+
+    my $matched = pf::authentication::match2($source_id, $merged, $$extra, \$attributes);
+
+    my $value = $matched->{values}{$Actions::TRIGGER_RADIUS_MFA} if $matched;
+    if ($value) {
+        my $mfa = pf::factory::mfa->new($value);
+        # If the mfa method is secondary password do nothing, the mfa will be triggered on a second request.
+        if ($mfa->radius_mfa_method eq 'second-password') {
+            my $chi = pf::CHI->new(namespace => 'mfa');
```

Comment:
switch this to `pf::mfa->cache` (that needs to be created based on a comment above)